torch/optim/adam.py

```diff
@@ -473,42 +477,51 @@ def _multi_tensor_adam(params: List[Tensor],
         torch._foreach_mul_(device_exp_avg_sqs, beta2)
         torch._foreach_addcmul_(device_exp_avg_sqs, device_grads, device_grads, 1 - beta2)
 
+        # Delete the local intermediate since it won't be used anymore to save on peak memory
+        del device_grads
+
         if capturable:
             bias_correction1 = torch._foreach_pow(beta1, device_state_steps)
             bias_correction2 = torch._foreach_pow(beta2, device_state_steps)
             # foreach_sub doesn't allow a scalar as the first arg
             torch._foreach_sub_(bias_correction1, 1)
             torch._foreach_sub_(bias_correction2, 1)
-            torch._foreach_neg_(bias_correction1)
+            # we do not negate bias_correction1 as it'll need to be negated later anyway
             torch._foreach_neg_(bias_correction2)
 
             # foreach_div doesn't allow a scalar as the first arg
-            step_size = torch._foreach_div(bias_correction1, lr)
-            torch._foreach_reciprocal_(step_size)
-            torch._foreach_neg_(step_size)
+            torch._foreach_div_(bias_correction1, lr)
+            torch._foreach_reciprocal_(bias_correction1)
 
-            bias_correction2_sqrt = torch._foreach_sqrt(bias_correction2)
+            torch._foreach_sqrt_(bias_correction2)
+
+            # Re-assign for clarity as we maintain minimal intermediates: we'll have
+            # step_size = - lr / (1 - beta1 ^ t) where t = num_steps
+            # bias_correction2 = 1 - beta2 ^ t
+            step_size = bias_correction1
```

Comment:
That sounds good. Just mentioning as well that you can do things like `step_size = torch._foreach_reciprocal_(bias_correction1)` or `bias_correction2_sqrt = torch._foreach_sqrt_(bias_correction2)` if you prefer.

Reply:
Nah, I'm p sure those return nothing. I had something like that before and they returned None. 
![image](https://github.com/pytorch/pytorch/assets/31798555/e95cf25b-fa5f-45a3-bebf-b9fb6f4df30e)

Reply:
What? :o  All our other inplace ops return the modified arg... This is very weird...

Reply:
"foreach support is beta" i guess? I can make another issue tracking this when we wanna make foreach legit

Reply:
just so you know IIRC the current torchgen doesn't allow inplace foreach functions to return a list of Tensors (I don't remember the exact lines off the top of my head)

Reply:
ioh that would make sense

Reply:
I re-grepped it:
https://github.com/pytorch/pytorch/blob/d8cb80e3827226eeaafd6a007018112301f89d41/torchgen/model.py#L1411-L1428